rfcs/0009-abstraction-layer-for-mobile-error-tracking.md

```diff
@@ -0,0 +1,191 @@
+} catch (err) {
+  errorTracking.captureError(err);
+}
+```
+
+# Motivation
+
+Let's suppose we were using [Sentry](https://sentry.io) as our solution. So, we would be able to capture an error using Sentry's `captureException` method.
+
+```js
+try {
+  aFunctionThatMightFail();
+} catch (err) {
+  Sentry.captureException(err);
+}
+```
+
+However we might decide to use another error tracking solution later, e.g. [Firebase Crashlytics](https://firebase.google.com/products/crashlytics). By doing so we would need to go through all our codebase and change all Sentry's methods to their Crashlytics' counterparts, which would be an unnecessary work.
+
+```js
+try {
+  aFunctionThatMightFail();
+} catch (err) {
+  crashlytics().recordError(err);
+}
+```
+
+What we would like to to instead is mapping our own API to the one from our current error tracking solution. Basically we would only need to import another module.
+
+Another important motivation is to add the ability to toggle the error reporting whenever we want to. Unfortunately, some solutions don't have this feature built-in, so we will create a way to handle this inside the abstraction layer.
+
+# Detailed design
+
+Example for Sentry:
+
+```js
+import * as Sentry from "@sentry/react-native";
+
+const captureError = Sentry.captureException;
+
+// level should be: fatal, error, warning, info, or debug.
+const addBreadcrumb = ({ scope, message, level = "info" }) =>
+  Sentry.addBreadcrumb({
+    category: scope,
+    message,
+    level,
+  });
+
+const setUserId = (level) =>
+  Sentry.configureScope((scope) => scope.setLevel(level));
+
+// You can set the severity of an event to one of five values: fatal, error, warning, info, and debug. error is the default, fatal is the most severe, and debug is the least severe.
+
+const setLevel = (id) =>
+  Sentry.configureScope((scope) => scope.setUser({ id }));
+
+const enable = (enabled) => {
+  isTrackingEnabled = enabled;
+};
+
+const createErrorTracking = ({ enabled = false } = {}) => {
+  let isTrackingEnabled = enabled;
+
+  const createTrackingFn = (fn) => (...args) => {
+    if (!isTrackingEnabled) {
+      return Promise.resolve();
+    }
+
+    return fn(...args);
+  };
+
+  return {
+    captureError: createTrackingFn(captureError),
+    addBreadcrumb: createTrackingFn(addBreadcrumb),
+    setUserId: createTrackingFn(setUserId),
+    setLevel: createTrackingFn(setLevel),
+    enable,
+  };
+};
+
+export default createErrorTracking;
+```
+
+Example for Firebase Crashlytics:
+
+```js
+import crashlytics from "@react-native-firebase/crashlytics";
+
+const captureError = crashlytics().recordError;
+
+const setUserId = crashlytics().setUserId;
+
+const setLevel = (level) => crashlytics().setAttributes({ level });
+
+// level should be: fatal, error, warning, info, or debug.
+const addBreadcrumbs = ({ scope, message, level = "info" }) =>
+  crashlytics().setAttributes({ category: scope, message, level });
+
+const enable = (enabled) => {
```

Comment:
We must reference `isTrackingEnabled` inside the factory function `createErrorTracking`. Otherwise, we cannot leverage the closure behavior.